RecoBTag/DeepFlavour/interface/TrackInfoBuilder.h

```diff
@@ -0,0 +1,119 @@
+#include "TrackingTools/IPTools/interface/IPTools.h"
+#include "TrackingTools/Records/interface/TransientTrackRecord.h"
+#include "RecoVertex/VertexTools/interface/VertexDistance3D.h"
+#include "TVector3.h"
+
+namespace deep{
+
+// adapted from DeepNtuples
+class TrackInfoBuilder{
+public:
+    TrackInfoBuilder(edm::ESHandle<TransientTrackBuilder> & build):
+        builder(build),
+        trackMomentum_(0),
+        trackEta_(0),
+        trackEtaRel_(0),
+        trackPtRel_(0),
+        trackPPar_(0),
+        trackDeltaR_(0),
+        trackPtRatio_(0),
+        trackPParRatio_(0),
+        trackSip2dVal_(0),
+        trackSip2dSig_(0),
+        trackSip3dVal_(0),
+        trackSip3dSig_(0),
+
+        trackJetDistVal_(0),
+        trackJetDistSig_(0)
+{
+
+
+}
+
+    void buildTrackInfo(const reco::Candidate * candidate ,const math::XYZVector&  jetDir, GlobalVector refjetdirection, const reco::Vertex & pv){
+        
+        // deal with PAT/AOD polymorphism to get track
+        const reco::Track * track_ptr = nullptr;
+        auto packed_candidate = dynamic_cast<const pat::PackedCandidate *>(candidate);
+        auto pf_candidate = dynamic_cast<const reco::PFCandidate *>(candidate);
+        if (pf_candidate) {
+          track_ptr = pf_candidate->bestTrack(); // trackRef was sometimes null
+        } else if (packed_candidate) {
+          track_ptr = &(packed_candidate->pseudoTrack());
+        }
+
+        reco::TransientTrack transientTrack;
+        transientTrack=builder->build(*track_ptr);
+        Measurement1D meas_ip2d=IPTools::signedTransverseImpactParameter(transientTrack, refjetdirection, pv).second;
+        Measurement1D meas_ip3d=IPTools::signedImpactParameter3D(transientTrack, refjetdirection, pv).second;
+        Measurement1D jetdist=IPTools::jetTrackDistance(transientTrack, refjetdirection, pv).second;
+        math::XYZVector trackMom = track_ptr->momentum();
+        double trackMag = std::sqrt(trackMom.Mag2());
+        TVector3 trackMom3(trackMom.x(),trackMom.y(),trackMom.z());
+        TVector3 jetDir3(jetDir.x(),jetDir.y(),jetDir.z());
+
+
+        trackMomentum_=std::sqrt(trackMom.Mag2());
+        trackEta_= trackMom.Eta();
+        trackEtaRel_=reco::btau::etaRel(jetDir, trackMom);
+        trackPtRel_=trackMom3.Perp(jetDir3);
+        trackPPar_=jetDir.Dot(trackMom);
+        trackDeltaR_=reco::deltaR(trackMom, jetDir);
+        trackPtRatio_=trackMom3.Perp(jetDir3) / trackMag;
+        trackPParRatio_=jetDir.Dot(trackMom) / trackMag;
+        trackSip2dVal_=static_cast<float>(meas_ip2d.value());
+
+        trackSip2dSig_=static_cast<float>(meas_ip2d.significance());
+        trackSip3dVal_=static_cast<float>(meas_ip3d.value());
+
+
+        trackSip3dSig_=static_cast<float>(meas_ip3d.significance());
+        trackJetDistVal_=static_cast<float>(jetdist.value());
+        trackJetDistSig_=static_cast<float>(jetdist.significance());
+
+    }
+
+    const float& getTrackDeltaR() const {return trackDeltaR_;}
```

Comment:
is ```const float&``` enforced by some design?
it is likely cheaper to have plain ```float``` return type.

Reply:
agreed, will change

Reply:
fixed with https://github.com/cms-sw/cmssw/pull/19893/commits/440b68d0ef3fee705cfd18936f7ad76d410c2471